content/drafts/using-sti-to-evolve-entities.md

```diff
@@ -0,0 +1,278 @@
+
+3. Have the "Task v1" and "Task v2" features know as little about the other as possible, and
+
+4. Have rest of the system's `Task`-adjacent features be largely agnostic about the v1/v2 dichotomy.
+
+And, of course, minimize both our short-term implementation and long-term maintenance pain.
+
+## Auditing Existing Usage
+
+Our first step was to understand how `Task` was used in the system, to understand the impact of our proposed changes.
+
+Given our system is a monolith, this was actually pretty easy to do: all incoming and outgoing connections (i.e. foreign keys from `tasks` to other tables, and foreign keys from other tables to `tasks`) to `Task` are listed in Joist's `TaskOpts` type, which a small sample looks like:
+
+```ts
+export interface TaskOpts {
+  // All primitive columns & enums...
+  durationInDays: number;
+  startDate: Date;
+  isStartNode?: boolean | null;
+  baselineStartDate?: Date | null;
+  internalNote?: string | null;
+  
+  // Also any "outgoing" FKs to other parts of the system
+  internalUser?: InternalUser | InternalUserId | null;
+  tradePartner?: TradePartner | TradePartnerId | null;
+  invoice?: Invoice | null;
+  
+  // Also well as any "incoming" FKs that depend on task
+  billLineItems?: BillLineItem[];
+  homeownerContractDraws?: HomeownerContractDraw[];
+  requiredTaskDocuments?: RequiredTaskDocument[];
+  documents?: Document[];
+  jobLogNotes?: JobLogNote[];
+}
+```
+
+Joist also produces a "reactivity report" that further shows how `Task` is connected to the rest of the system, beyond the immediate foreign keys, via Joist's reactive rules & reactive fields features:
+
+```txt
+Task will trigger reactions in:
+  HomeownerNote
+    [homeownerVisible] -> homeownerNotes HomeownerNote.ts:127
+  RequiredTaskDocument
+    [documents] -> requiredTaskDocuments RequiredTaskDocument.ts:8
+  ScheduleSubPhase
+    [scheduleSubPhase,endDate,globalTask] -> scheduleSubPhase targetEndDate
+```
+
+Somewhat surprisingly, `Task` did not have a meaningful amount of reactivity, which was actually good news for our purposes!
+
+So, just looking at `TaskOpts`, in total we had:
+
+* 23 columns
+* 7 outgoing relations
+* 17 incoming relations
+
+Which in total was about 50 fields/relations to consider. We then made notes on each of "what is staying vs. being deprecated", i.e.:
+
+<div style="font-style: italic; display: flex; flex-direction: column; align-items: center">
+  <img src="/images/sti/task-audit.png" style="width:50em;"/>
+  Audit of our Task relations
+</div>
+
+This audit was great because it gave us context & concrete data points for then evaluating our high-level approaches, instead of just our initial gut feelings. 
+
+## High Level Approaches
+
+We generally had three approaches to consider:
+
+1. Create an entirely new table/entity, like `tasks_v2` / `TasksV2`, or
+2. Reuse the existing `tasks` / `Task` table/entity, and somehow manage the v1/v2 differentiation
```

Comment:
```suggestion
2. Reuse the existing `tasks` / `Task` table/entity, and somehow manage the v2/v3 differentiation
```
maybe not an important distinction, but you do point out above that this is v3?

Reply:
Ehh I see there are more mentions of v2 below, maybe simpler to hope the reader doesn't care about that distinction

Reply:
Yep...I should probably just pretend that "the real v3" is "just v2" for the purposes of the blog post.